reducer.py

```diff
@@ -0,0 +1,319 @@
+
+    @abc.abstractmethod
+    def reduce(metrics):
+        """User provides this."""
+        pass
+
+
+class MyAverageSimpleReducer(det.SimpleReducer):
+    """This is about as simple as a user-defined custom reducer can be."""
+    def reduce(metrics):
+        return sum(metrics) / len(metrics)
+
+
+# How would we integrate with EstimatorTrial?
+# We would integrate via custom SessionRunHooks. Something like this:
+
+class EstimatorTrialContext:
+    ...
+    def allreduce_metrics(self, reducer, metrics):
+        """
+        Why expose this as part of the context?  We have customers asking for the ability
+        to see metrics in tensorboard which are being calculated by hand in EvalHooks, but which
+        are not being shared across GPUs, so their tensorboard metrics are not quite right.
+
+        Arguably, in those cases we don't need a two-step reducer like det.Reducer would be;
+        they would be able to do the per_slot_reduce() on their own in their EvalHook.
+
+        Frankly, I'm not convinced that this should be exposed to users; it seems like they could
+        skip implementing their custom EvalHook with the reducer interface.
+        """
+
+        this_slot_metrics = reducer.per_slot_reduce(metrics)
+
+        # Handle the single-slot case.
+        if self.distributed.size() == 0:
+            return reducer.cross_slot_reduce(this_slot_metrics)
+
+        # Handle the chief case.
+        if self.distributed.rank() == 0:
+            per_slot_metrics = chief_gather_per_slot_metrics(this_slot_metrics)
+            reduced = self.reducer.cross_slot_reduce(per_slot_metrics)
+            chief_distribute_reduce_metrics(reduced)
+        else:
+            reduced = worker_allreduce_metrics(this_slot_metrics)
+        return reduced
+
+
+class ReducerEvalHook(tf.compat.v1.train.SessionRunHook):
+    """
+    Apply the reducer to the outputs of the fetches from each batch, and write the final result to
+    metric_name in tensorboard.
+
+    Users would include an instance of ReducerEvalHook as part of their EvalSpec.
+    """
+
+    def __init__(self, context, fetches, reducer, metric_name):
+        self.context = context
+        self.fetches = fetches
+        self.reducer = reducer
+        self.metric_name = metric_name
+
+    def begin(self):
+        self.batch_metrics = []
+
+    def before_run(self, run_contex):
+        return tf.estimator.SessionRunArgs(fetches=self.fetches)
+
+    def after_run(self, run_contex, run_values):
+        self.per_slot_metrics.append(run_values.results)
+
+    def end(self, session):
+        reduced = self.context.allreduce_metrics(self.reducer, self.batch_metrics)
+
+        write_metric_to_tensorboard(self.metric_name, reducer)
+
+
+class MyLinearEstimator(estimator.EstimatorTrial):
+    ...
+
+    def build_validation_spec(self) -> tf.estimator.EvalSpec:
+        def fn():
+            ...
+
+        global_step_metric_hook = det.estimator.ReducerEvalHook(
```

Comment:
It's not very clear to me how we will be able to support evaluation metrics (not metrics in EvalHooks) via this interface.

Reply:
Yeah, this interface is not really intended to support the tf.metrics that are actually in the graph at all.